lib/charms/opensearch/v0/opensearch_base_charm.py

```diff
@@ -244,6 +244,7 @@ def __init__(self, *args, distro: Type[OpenSearchDistribution] = None):
             metrics_endpoints=[],
             scrape_configs=self._scrape_config,
             refresh_events=[
+                self.on.config_changed,
```

Comment:
The `COSAgentProvider` has [self.on.config_changed](https://github.com/canonical/grafana-agent-operator/blob/main/lib/charms/grafana_agent/v0/cos_agent.py#L655C81-L656C1) by default if not custom events  are passed.

We are passing custom events, no there is no refresh on COS for config-changed.
When you upgrade a charm, after that the config-change hook is triggered.

In summary, I'm re adding the default behavior and the same time cover my need of refreshing COS relation after charm upgrade

Reply:
Another thing...
I didn't add `self.on_charm_upgrade` because I was seeing a race condition. Talking with the [COS team](https://matrix.to/#/!nHXpRkcSNJHlHdUGbQ:ubuntu.com/$ravfUK4XN39YUeqybKqMiDBzuPo0fzh9Y_WJp3sVhbs?via=ubuntu.com&via=matrix.org), the order when you add those handlers matters and COS Provider should in theory be added as a last handler.

To do that, we would need to move the `COSAgentProvider` to the last child class of the inheritance and I though that it would be simpler just add back the config-changed

Reply:
Shouldn't we refresh on COS broken event as well? It is a slightly off topic question for this PR but how can we assure the dashboard is gone if we remove the COS relation from opensearch?

Reply:
I've tested and as expected (at least for me), the dashboards are removed if the relation with grafana-agent is removed